src/cli/expand-patterns.js

```diff
@@ -0,0 +1,234 @@
+"use strict";
+
+const path = require("path");
+const fs = require("fs");
+const globby = require("globby");
+const fastGlob = require("fast-glob");
+const flat = require("lodash/flatten");
+
+/** @typedef {import('./util').Context} Context */
+
+/**
+ * @param {Context} context
+ */
+function* expandPatterns(context) {
+  const cwd = process.cwd();
+  const seen = Object.create(null);
+  let noResults = true;
+
+  for (const pathOrError of expandPatternsInternal(context)) {
+    noResults = false;
+    if (typeof pathOrError !== "string") {
+      yield pathOrError;
+      continue;
+    }
+
+    const relativePath = path.relative(cwd, pathOrError);
+
+    // filter out duplicates
+    if (relativePath in seen) {
+      continue;
+    }
+
+    seen[relativePath] = true;
+    yield relativePath;
+  }
+
+  if (noResults) {
+    // If there was no files and no other errors, let's yield a general error.
+    yield {
+      error: `No matching files. Patterns: ${context.filePatterns.join(" ")}`
+    };
+  }
+}
+
+// TODO: use `fast-glob` directly, without `globby`
+const baseGlobbyOptions = {
+  dot: true,
+  expandDirectories: false,
+  absolute: true
+};
+
+/**
+ * @param {Context} context
+ */
+function* expandPatternsInternal(context) {
+  // Ignores files in version control systems directories and `node_modules`
+  const ignoredDirectories = {
+    ".git": true,
+    ".svn": true,
+    ".hg": true,
+    node_modules: context.argv["with-node-modules"] !== true
+  };
+
+  const globbyOptions = {
+    ...baseGlobbyOptions,
+    ignore: Object.keys(ignoredDirectories)
+      .filter(dir => ignoredDirectories[dir])
+      .map(dir => "**/" + dir)
+  };
+
+  let supportedFilesGlob;
+  const cwd = process.cwd();
+
+  /**
+   * @type {Array<{
+   *  type: 'file' | 'dir' | 'glob';
+   *  absolutePath?: string;
+   *  glob?: string;
+   *  input: string;
+   * }>}
+   */
+  const entries = [];
+
+  for (const pattern of context.filePatterns) {
+    const absolutePath = path.resolve(cwd, pattern);
+
+    if (containsIgnoredPathSegment(absolutePath, cwd, ignoredDirectories)) {
```

Comment:
What about using `micromatch` here, we already have it in our deps (from `fast-glob`)

Reply:
This code is only for silent ignoring files in `node_modules` and VCS dirs. I think it's good enough.